### ISSUE TITLE
Handle recurring revenue

### ISSUE THREAD

**Issue by Blackclaws** [JC5/firefly-iii#359]
So I was wondering whether there is any way to handle recurring income such as monthly paychecks that don't change every month. 

I only found a way to define monthly expenses.

Or is the preferred way to always import the exports from my bank?

**Comment by JC5** [comment]
I wish to implement recurring withdrawals, transfers and deposits. It is a
very good idea. To create these might be complicated, so it will take some
time.

These would not be equal to "bills", but something new I have yet to think
about.

Thank you for your idea! I will update this issue when I have something to
show you.

**Comment by JC5** [comment]
Some general notes.

The object that holds repetition info is called a **repeated_transaction** in this text.

Like Outlook / GCal a button / option / checkbox to repeat a transaction. Repeat settings show in popup (modal dialog) which I will probably steal from Google.

Options for point 2 and point 3 are radiobox.

1. Every X days / weeks / months / years (dropdown)
2. Specify which day:
  1. on this day of week: (if week) mo tu we th fr sa su (checkbox)
  2. on this day of month: (if month) 1, 2, 3, 4 (select list)
  3. 3rd wednesday of month (if month)
  4. (no options when user picks year or day)
3. Repeats until
  1. infinite
  2. x times (input)
  3. until date (input)
4. summary in readable line.

Then, dialog closes and readable line is shown somehwere in add transaction box to indicate the transaction will repeat, starting on the date of the transaction.

When transaction is created, repeated_transaction object. is too. These can be edited on separate page (not on transaction edit page). 

So need a separate set of pages for repeated_transaction [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete).

Repeated_transaction is a separate object with all data like a transaction (description, budget, amount, etc) plus the repetition info. Whenever the repetition hits (cron-job or user interface call?) transaction is created that links to repeated_transaction thing.

Edits to repeated_transaction do NOT edit past transactions, only future. Use the mass edit to edit past transactions. Link to repeated_transaction is never lost, even when repeated_transaction is edited.

repeated_transaction can also be created separately in a dedicated form. Will be the same as modal (but not in a modal) + all the data a transaction needs.

Need to think about meta info such as dates, especially rent date and processing date. Lets say you change processing date to one day later. Will future repetitions do this also?

Any field missing from the repeated_transaction will be missing from the journal but system will try to create the transaction anyway. This is a moment to start using laravel's notification thing.

Match with rules and match with bills will be automated when transaction is actually created.

**Comment by nekromoff** [comment]
I suggest that all the expense data are pulled from outside channels such as bank exports or by manual addition rather than adding such complexity to the system.

**Comment by dreeco** [comment]
Hello, first of all: amazing piece of software you have here!! Congrats.
Why are regular revenues different from bills? Implementing a negative bill amount could do the job for a first implementation, nope ?
Bills could benefit from a begin / end date as well. Some bills (loan reimbursement from example) are painful to enter x times.

Adrien

**Comment by mjhd** [comment]
This seems like an absolutely critical feature, I really need to implement the negative bill - I'm just gonna do it myself...

I spoke with my mom, who is a lifelong bookkeeper, and she seemed pretty certain that there is no technical difference between an expense and an income. She put it like this: if you would pay a bill on the first of the month, the company you pay it to would record the EXACT SAME transaction, except from their point of view it is income. What happens after that may be different, but both metrics measure the same transaction, just in a different directions. So I feel pretty confident this is a valid approach. Unfortunately if my paycheck doesn't show in my dashboard, it's not very useful to me, I can't really track my finances otherwise. Awesome software though - light-years better than everything else I've tried

**Comment by JC5** [comment]
Well it's been **more than two years** since this issue was closed, so in the mean time I built recurring transactions that you can use.

**Comment by mjhd** [comment]
Ok interesting, I tried that actually, but thought I didn't understand - when I enter one it isn't reflected on the day I schedule it for on the home accounts graph like a normal deposit is, so I thought I just didn't understand. Do I have to change a setting for it to be shown?

**Comment by JC5** [comment]
No, the function allows you to plan a deposit. When it fires, which requires the cron job to be active, it will generate a standard deposit like all others, which should be visible everywhere.

**Comment by JC5** [comment]
Hi! This is a canned reply that tells you how to get more debug details.

1. Please collect some log files. Go to `storage/logs` and open the latest log file. It should be called something like "ff3-fpm-fcgi-*.log". In it, you should be able to find the error along with its "stacktrace": this will tell me where Firefly III broke down. If you run Docker, use `docker logs <container`.
2. You may also [enable debug mode](https://github.com/firefly-iii/help/wiki/Enable-debug-logging-and-debug-mode) to make sure the log files catch everything and try again.
3. Make sure you clear your browser cookies and clear Firefly III's cache files by running `php artisan cache:clear` or by visiting `/flush` on your installation.

If you do not want to share log files or other details on Github, please [send them to me over (encrypted) email](https://firefly-iii.org/contact-information.html).

**Comment by viharm** [comment]
I'm using Firefly III v4.8.2, and trying to setup a recurring deposit, but the expense accounts are not showing up in the dropdown with the list of source accounts.
![image](https://user-images.githubusercontent.com/2663937/71783534-f69e3f80-2fdf-11ea-8870-270214d6a56f.png)


Is this intentional? If so, is there a way around it?

**Comment by JC5** [comment]
Select “Deposit” under “Transaction type”.

**Comment by viharm** [comment]
Oops, apologies for the oversight. But thanks for the super-prompt response :-)
